ui/shared/AssetAndAmountInput.qml

```diff
@@ -0,0 +1,134 @@
+import QtQuick 2.13
+import QtQuick.Controls 2.13
+import QtQuick.Layouts 1.13
+import QtGraphicalEffects 1.13
+import "../imports"
+
+Item {
+  id: root
+  property var assets
+  property string errorMessage: ""
+  property alias text: inputAmount.text
+  signal modelChanged(value: string)
+  height: inputAmount.height + txtFiatBalance.height + txtFiatBalance.anchors.topMargin
+  anchors.right: parent.right
+  anchors.left: parent.left
+
+
+  Item {
+    id: txtBalanceInfo
+    anchors.right: parent.right
+    anchors.left: parent.left
+    anchors.top: parent.top
+    anchors.topMargin: 0
+    height: txtBalanceDesc.height
+
+    StyledText {
+        id: txtBalanceDesc
+        text: qsTr("Balance: ")
+        anchors.right: txtBalance.left
+        font.weight: Font.Medium
+        font.pixelSize: 13
+        color: Style.current.grey
+    }
+
+    StyledText {
+        id: txtBalance
+        text: selectAsset.selectedAsset.value
+        anchors.right: parent.right
+        font.weight: Font.Medium
+        font.pixelSize: 13
+        color: Style.current.grey
+
+        MouseArea {
+            cursorShape: Qt.PointingHandCursor
+            anchors.fill: parent
+            hoverEnabled: true
+            onExited: {
+                txtBalance.color = Style.current.grey
+            }
+            onEntered: {
+                txtBalance.color = Style.current.black
+            }
+            onClicked: {
+                inputAmount.setModel(selectAsset.selectedAsset.value)
+            }
+        }
+    }
+  }
+
+  Input {
+      id: inputAmount
+      label: qsTr("Asset & Amount")
+      placeholderText: "0.00"
+      validationError: root.errorMessage
+      anchors.top: parent.top
+      customHeight: 56
+      onTextChanged: {
+        root.modelChanged(inputAmount.text)
+      }
+      onModelChanged: {
+        let amount = inputAmount.text.trim()
+        if (amount === "" || isNaN(amount)) {
+          return
+        }
+        txtFiatBalance.setModel(walletModel.getFiatValue(amount, selectAsset.selectedAsset.symbol, "USD"))
+      }
```

Comment:
I've introduce `onModelChanged` because we're implementing a two-way data binding here that otherwise leads to infinite call stack growth..

Instead of reacting to `onTextChanged` and then updating the `fiatBalance.text`, which then triggers its `onTextChanged`, which then changes `inputAmount.text`, which then again triggers its `onTextChanged`..., we're using a custom `onModelChanged` signal and a `setModel()` interceptor to control when `onModelChanged` is emitted (e.g. it shouldn't emit from a text change, when the text change was triggered by a `setModel`)

If there's a better QML way of doing this, I'm happy to change this.

Reply:
Also, discussed with @emizzle that we probably want to make these handlers configurable instead of relying on global `walletModel` here

Reply:
One solution would be to not use `onTextChanged` but instead use:
```qml
Keys.onReleased: {
      // Check the new text now
}
```
Sorry if I didn't suggest it yesterday, I forgot about it

Reply:
Where is `modelChanged` being fired/consumed? Maybe it's leftover?

Do you think that `AssetAndAmountInput` should not depend on `walletModel` APIs (similar to what we discussed in DMs)? Would it be possible that every time this component calls `walletModel.getFiatValue`, instead trigger a signal that the *calling* component consumes and sends off to the `walletModel` APIs. This would then update an internal property that all internal text components that depend on the updated value would use to display the fiat value. Something like this:

```qml
// AssetAndAmountInput.qml
Item {
    id: root
    signal inputChanged(string input, string symbol, string currency) // called every time the input has changed or fiat value needs to be updated
    property string fiatBalanceDisplay = "0.00" // all internal components use this property to display the fiat valiue
    
    Input {
      id: inputAmount
      Keys.onReleased: {
          let amount = inputAmount.text.trim()
          if (amount === "" || isNaN(amount)) {
              return
          }
          root.inputChanged(amount, selectAsset.selectedAsset.symbol, walletModel.defaultCurrency)
      }

    StyledTextField {
          id: txtFiatBalance
          text: root.fiatBalanceDisplay
    }
    // Repeat for all other instances of input changing...
}

// Caller.qml
AssetAndAmountInput {
    assets: walletModel.assets
    onInputChanged: {
        this.fiatBalanceDisplay = walletModel.getFiatValue(input, symbol, currency)
    }
}
```

Reply:
So to be very honest, I don't necessarily think it'd be good if the caller has to take of providing the functionality of the component that she uses. E.g. the job of the AssetAndAmountInput is not only to hold an amount + its selected asset and also emitting said data in case it changes, it also needs to take care of rendering the proper fiat balance and update the crypto amount, in case the fiat balance changes.

Moving this out into consumer APIs feels a little backwards to me because now, the component practically doesn't work, if the caller doesn't provide the core functionality of the component. To give a better example, we also don't ask the caller to update the **internal** AssetSelector of the amount input, when the account has changed. We let the component take care of it.

I do realize however that the same or similar issue exists when we just rely on walletModel APIs to exist inside of that component. Ideally, what we want, is that the component comes with that functionality.
I see 2 options to make that possible:

1. We either prepare a Qt View class for the AssetAndAmountSelector, which is simply part of the corresponding QML and also takes care of supplying the UI with the correct logic. That still requires the caller to instantiate said Qt object and bind it to the QML instance, however user won't be required to implement the functionality

2. We implement the functionality in QML and JavaScript, which is most likely possible as well. In that case we can remove Nim from the equation entirely, which would make the QML component even more reusable.

Anyways, went ahead and implemented your suggestion for now

Reply:
Very fair point. Do you think it would be possible to pass the API to the component via a property so it can use it internally?

```
AssetAndAmountSelector {
    getFiatValue: walletModel.getFiatValue
    convertValue: walletModel.convertValue
}
```

To prevent the component from not working without this, the default value for the properties could be a function that simply returns the input param (IOW, do no conversion at all).

If not, or if that’s an inferior suggestion, I agree with #2. It should be doable with a simple http request, however so js library help would be nice in this situation.

Reply:
> Do you think it would be possible to pass the API to the component via a property so it can use it internally

Yea, I think that should work to and is probably a nice compromise. Lemme give it a shot.
Would like to try the JS route in the future, but I'll refrain from experimenting with this now, so we can land the component.

Reply:
Okay so... seems like passing down a function like this will cause nim to crash with an infinite loop.. Same happens when the  functions are wrapped..

Reply:
Great job getting this fixed and glad to see it's working without relying on the `walletModel`! 🙌